cachi2/core/package_managers/yarn_classic/project.py

```diff
@@ -0,0 +1,173 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Any, Literal, NamedTuple, Union
+
+from pyarn import lockfile  # type: ignore
+
+from cachi2.core.errors import PackageRejected
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(name=__name__)
+
+ConfigKind = Literal["package_json", "yarnlock"]
+
+
+@dataclass
+class _CommonConfigFile(ABC):
+    """A base class for representing a config file.
+
+    :param path: the path to the config file, relative to the request source dir
+    :param data: the raw data for the config file content
+    """
+
+    _path: RootedPath
+    _data: dict[str, Any]
+
+    @property
+    def data(self) -> dict[str, Any]:
+        return self._data
+
+    @property
+    def path(self) -> RootedPath:
+        return self._path
+
+    @property
+    @abstractmethod
+    def config_kind(self) -> ConfigKind:
+        """Return kind of ConfigFile instance."""
+        pass
+
+    @classmethod
+    @abstractmethod
+    def from_file(cls, path: RootedPath) -> "_CommonConfigFile":
+        """Construct a ConfigFile instance."""
+        pass
+
+
+class PackageJson(_CommonConfigFile):
+    """A package.json file.
+
+    This class abstracts the underlying attributes and only exposes what
+    is relevant for the request processing.
+    """
+
+    @property
+    def config_kind(self) -> ConfigKind:
+        """Return kind of this ConfigFile."""
+        return "package_json"
+
+    @property
+    def install_config(self) -> dict[str, Any]:
+        """Get the installConfig dict."""
+        return self.data.get("installConfig", {})
+
+    @classmethod
+    def from_file(cls, path: RootedPath) -> "PackageJson":
+        """Construct a PackageJson instance."""
+        try:
+            package_json_data = json.loads(path.path.read_text())
+        except FileNotFoundError:
+            raise PackageRejected(
+                reason="The package.json file must be present for the yarn package manager",
+                solution=(
+                    "Please double-check that you have specified the correct path "
+                    "to the package directory containing this file"
+                ),
+            )
+        except json.decoder.JSONDecodeError as e:
+            raise PackageRejected(
+                reason=f"Can't parse the {path.subpath_from_root} file. {e}",
+                solution=(
+                    "The package.json file must contain valid JSON. "
+                    "Refer to the parser error and fix the contents of the file."
+                ),
+            )
+
+        return cls(path, package_json_data)
+
+
+class YarnLock(_CommonConfigFile):
+    """A yarn.lock file.
+
+    This class abstracts the underlying attributes.
+    """
+
+    yarn_lockfile: lockfile.Lockfile
+
+    @property
+    def config_kind(self) -> ConfigKind:
+        """Return kind of this ConfigFile."""
+        return "yarnlock"
```

Comment:
IMO, this property should be dropped and replaced with `isinstance(<class-name>)` somewhere else in the code.

Reply:
The use case here is being able to use the pattern `what_are_you = foo.config_kind` rather than being forced to use

```
if isinstance(foo):
    what_are_you = i_am_foo
elif isinstance(bar):
    what_are_you = i_am_bar
elif isinstance(baz):
    what_are_you = i_am_baz
```

every time.

If we're truly switching a different code block for *each* and *all* classes, then yeah, the if-tree is necessary.

Reply:
What would be the real use case here ? `__CommonConfigFile` is the abstract class so you can't instantiate it anyway.
`kind = PackageJson.config_kind == "package_json"` this will awkward to me.

Reply:
EDIT: So my point about `isinstance` is not valid, `ConfigKind` is just Literal.

Reply:
Exactly :)

Reply:
But still don't understand the use case of such property.

Reply:
EDIT 2: I understand, but using the class as type insted of literal is much clearer IMO.